packages/cli/src/services/frontend.service.ts

```diff
@@ -0,0 +1,70 @@
+import { Service } from 'typedi';
+import uniq from 'lodash/uniq';
+import { createWriteStream } from 'fs';
+import { mkdir } from 'fs/promises';
+import path from 'path';
+
+import type { ICredentialType, INodeTypeBaseDescription } from 'n8n-workflow';
+
+import { GENERATED_STATIC_DIR } from '@/constants';
+import { CredentialsOverwrites } from '@/CredentialsOverwrites';
+import { CredentialTypes } from '@/CredentialTypes';
+import { LoadNodesAndCredentials } from '@/LoadNodesAndCredentials';
+
+@Service()
+export class FrontendService {
+	constructor(
+		private readonly loadNodesAndCredentials: LoadNodesAndCredentials,
+		private readonly credentialTypes: CredentialTypes,
+		private readonly credentialsOverwrites: CredentialsOverwrites,
+	) {}
+
+	async generateTypes() {
+		this.overwriteCredentialsProperties();
+
+		// pre-render all the node and credential types as static json files
+		await mkdir(path.join(GENERATED_STATIC_DIR, 'types'), { recursive: true });
+		const { credentials, nodes } = this.loadNodesAndCredentials.types;
+		this.writeStaticJSON('nodes', nodes);
+		this.writeStaticJSON('credentials', credentials);
+	}
+
+	private writeStaticJSON<T extends INodeTypeBaseDescription | ICredentialType>(
+		name: string,
+		data: T[],
+	) {
```

Comment:
This method does not need a generic.

https://effectivetypescript.com/2020/08/12/generics-golden-rule/